core/src/main/java/jenkins/security/RSAConfidentialKey.java

```diff
@@ -0,0 +1,88 @@
+package jenkins.security;
+
+import org.apache.commons.codec.binary.Base64;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.KeyFactory;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.PrivateKey;
+import java.security.SecureRandom;
+import java.security.interfaces.RSAPrivateCrtKey;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.RSAPublicKeySpec;
+
+/**
+ * RSA public/private key pair as {@link ConfidentialKey}.
+ *
+ * <p>
+ * As per the design principle of {@link ConfidentialKey}, not exposing {@link PrivateKey} directly.
+ * Define subtypes for different use cases.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+public abstract class RSAConfidentialKey extends ConfidentialKey {
+    private volatile RSAPrivateKey priv;
+    private volatile RSAPublicKey pub;
+    public RSAConfidentialKey(String id) {
+        super(id);
+    }
+
+    public RSAConfidentialKey(Class owner, String shortName) {
+        this(owner.getName() + '.' + shortName);
+    }
+
+    private RSAPrivateKey getKey() {
```

Comment:
@kohsuke can I ask you to recall exactly what code pattern of yours offends me most?

So let me see, you use a volatile to avoid locking on read, which incurs a memory barrier, and then use a synchronised to guard the write (which costs the same memory barrier except now you have additional memory barriers on each access of the volatile field)

Unless you can prove the need for this (incorrect in my view) complexity, please just use a lock to guard read and write.

(I have yet to review the actual code, but you called out my number 1 alarm bell of threading issues in Jenkins, and I am not yet - from my phone - seeing any comment where you claim the need to use this pattern)

Actually scratch that. I've looked at the code. Please use a lock as the volatile incurs the same memory barrier cost for all reads as a simple lock and incurs far more memory barriers for a write and this is a relatively simple class. Use an internal lock if you want to avoid using the objects intrinsic lock (which is normally a good idea, don't want others enticing deadlocks by accidentally locking on the intrinsic lock)